jaeger-core/src/main/java/io/jaegertracing/internal/reporters/FilteringReporter.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2021, The Jaeger Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.jaegertracing.internal.reporters;
+
+import io.jaegertracing.internal.JaegerSpan;
+import io.jaegertracing.internal.JaegerSpanContext;
+import io.jaegertracing.internal.metrics.Metrics;
+import io.jaegertracing.spi.Reporter;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.LongAdder;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * FilteringReporter drops spans below a specific threshold and buffers the remaining spans in memory and
+ * sends them to a delegate only once a parent span exceeds a specified threshold.
+ */
+@ToString
+@Slf4j
+public class FilteringReporter implements Reporter {
+  public static final long DEFAULT_FILTER_SPANS_UNDER_MICROS = 0L;
+  public static final long DEFAULT_DEFER_SPANS_UNDER_MICROS = 0L;
+
+  private final Reporter delegate;
+  private final long filterSpansUnderMicros;
+  private final long deferSpansUnderMicros;
+  private final Metrics metrics;
+
+  private final Map<Long, List<JaegerSpan>> pendingByParent = new ConcurrentHashMap<>();
+  private final LongAdder pendingCount = new LongAdder();
+
+  public FilteringReporter(
+      Reporter delegate, long filterSpansUnderMicros, long deferSpansUnderMicros, Metrics metrics) {
+    this.delegate = delegate;
+    this.filterSpansUnderMicros = filterSpansUnderMicros;
+    this.deferSpansUnderMicros = deferSpansUnderMicros;
+    this.metrics = metrics;
+  }
+
+  @Override
+  public void report(JaegerSpan span) {
+    final JaegerSpanContext context = span.context();
+    final List<JaegerSpan> pendingChildren = pendingByParent.remove(context.getSpanId());
```

Comment:
The guarantees this code provides is _very_ much not clear from the code without playing it out in your head.  This definitely needs a doc comment explaining the potential states we can hit in the `defer` case so that others can replay this state machine in their heads without having to re-create it from scratch.